pkg/daemon/pinned_image_set.go

```diff
@@ -0,0 +1,1314 @@
+	runtimeEndpoint string
+	// timeout for the prefetch operation
+	prefetchTimeout time.Duration
+	// backoff configuration for retries.
+	backoff wait.Backoff
+
+	// cache for reusable image information
+	cache *imageCache
+
+	syncHandler              func(string) error
+	enqueueMachineConfigPool func(*mcfgv1.MachineConfigPool)
+	queue                    workqueue.RateLimitingInterface
+	featureGatesAccessor     featuregates.FeatureGateAccess
+
+	// mutex to protect the cancelFn
+	mu       sync.Mutex
+	cancelFn context.CancelFunc
+}
+
+// NewPinnedImageSetManager creates a new pinned image set manager.
+func NewPinnedImageSetManager(
+	nodeName string,
+	criClient *cri.Client,
+	mcfgClient mcfgclientset.Interface,
+	imageSetInformer mcfginformersv1alpha1.PinnedImageSetInformer,
+	nodeInformer coreinformersv1.NodeInformer,
+	mcpInformer mcfginformersv1.MachineConfigPoolInformer,
+	minStorageAvailableBytes resource.Quantity,
+	runtimeEndpoint string,
+	authFilePath string,
+	prefetchTimeout time.Duration,
+	featureGatesAccessor featuregates.FeatureGateAccess,
+) *PinnedImageSetManager {
+	p := &PinnedImageSetManager{
+		nodeName:                 nodeName,
+		mcfgClient:               mcfgClient,
+		runtimeEndpoint:          runtimeEndpoint,
+		authFilePath:             authFilePath,
+		prefetchTimeout:          prefetchTimeout,
+		minStorageAvailableBytes: minStorageAvailableBytes,
+		featureGatesAccessor:     featureGatesAccessor,
+		queue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "pinned-image-set-manager"),
+		prefetchCh:               make(chan prefetch, defaultPrefetchWorkers*2),
+		criClient:                criClient,
+		backoff: wait.Backoff{
+			Steps:    maxRetries,
+			Duration: retryDuration,
+			Factor:   retryFactor,
+			Cap:      retryCap,
+		},
+	}
+
+	imageSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.addPinnedImageSet,
+		UpdateFunc: p.updatePinnedImageSet,
+		DeleteFunc: p.deletePinnedImageSet,
+	})
+
+	mcpInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    p.addMachineConfigPool,
+		UpdateFunc: p.updateMachineConfigPool,
+		DeleteFunc: p.deleteMachineConfigPool,
+	})
+
+	p.syncHandler = p.sync
+	p.enqueueMachineConfigPool = p.enqueue
+
+	p.imageSetLister = imageSetInformer.Lister()
+	p.imageSetSynced = imageSetInformer.Informer().HasSynced
+
+	p.nodeLister = nodeInformer.Lister()
+	p.nodeListerSynced = nodeInformer.Informer().HasSynced
+
+	p.mcpLister = mcpInformer.Lister()
+	p.mcpSynced = mcpInformer.Informer().HasSynced
+
+	p.cache = newImageCache(256)
+
+	return p
+}
+
+func (p *PinnedImageSetManager) sync(key string) error {
+	klog.V(4).Infof("Syncing PinnedImageSet for pool: %s", key)
+	node, err := p.nodeLister.Get(p.nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to get node %q: %w", p.nodeName, err)
+	}
+
+	pools, _, err := helpers.GetPoolsForNode(p.mcpLister, node)
```

Comment:
I think this still doesn't fully match MachineConfig behaviour, but let's do that as a follow up once we merge this, if it's something we care about

Reply:
Discussed with Sam, I think we're covered here actually, apologies for the noise